File: server/embeddings.py
"""
Embedding service for generating vector embeddings.

Supports multiple providers:
- openai: OpenAI API (text-embedding-3-small, etc.)
- local: Local models via fastembed (BAAI/bge-small-en-v1.5, etc.)
"""

import os
import re
from typing import List, Optional, Literal
from abc import ABC, abstractmethod
import logging

logger = logging.getLogger(__name__)

# Type alias for supported providers
EmbeddingProvider = Literal["openai", "local"]


class BaseEmbeddingProvider(ABC):
    """Abstract base class for embedding providers."""

    @abstractmethod
    def generate_embedding(self, text: str) -> List[float]:
        """Generate embedding for a single text."""
        pass

    @abstractmethod
    def generate_embeddings_batch(self, texts: List[str]) -> List[List[float]]:
        """Generate embeddings for multiple texts."""
        pass

    @abstractmethod
    def get_dimension(self) -> int:
        """Get the embedding dimension."""
        pass


class OpenAIProvider(BaseEmbeddingProvider):
    """OpenAI API embedding provider."""

    def __init__(self, api_key: str, model: str = "text-embedding-3-small"):
        from openai import OpenAI
        self.client = OpenAI(api_key=api_key)
        self.model = model
        self._dimension = 1536  # Default for text-embedding-3-small
        logger.info(f"Initialized OpenAI provider with model: {model}")

    def generate_embedding(self, text: str) -> List[float]:
        response = self.client.embeddings.create(model=self.model, input=text)
        return response.data[0].embedding

    def generate_embeddings_batch(self, texts: List[str]) -> List[List[float]]:
        response = self.client.embeddings.create(model=self.model, input=texts)
        return [data.embedding for data in response.data]

    def get_dimension(self) -> int:
        return self._dimension


class LocalProvider(BaseEmbeddingProvider):
    """Local embedding provider using fastembed."""

    def __init__(self, model: str = "BAAI/bge-small-en-v1.5"):
        try:
            from fastembed import TextEmbedding
        except ImportError:
            raise ImportError(
                "fastembed is required for local embeddings. "
                "Install with: pip install fastembed"
            )
        self.model_name = model
        self.model = TextEmbedding(model_name=model)
        # bge-small-en-v1.5 has 384 dimensions
        self._dimension = 384
        logger.info(f"Initialized local provider with model: {model}")

    def generate_embedding(self, text: str) -> List[float]:
        embeddings = list(self.model.embed([text]))
        return embeddings[0].tolist()

    def generate_embeddings_batch(self, texts: List[str]) -> List[List[float]]:
        embeddings = list(self.model.embed(texts))
        return [emb.tolist() for emb in embeddings]

    def get_dimension(self) -> int:
        return self._dimension


class EmbeddingService:
    """Service for generating embeddings using configurable providers."""

    def __init__(
        self,
        provider: Optional[EmbeddingProvider] = None,
        api_key: Optional[str] = None,
        model: Optional[str] = None
    ):
        """
        Initialize the embedding service.

        Args:
            provider: Embedding provider ("openai" or "local").
                      Defaults to EMBEDDING_PROVIDER env var or "openai".
            api_key: OpenAI API key (only needed for openai provider).
                     Defaults to OPENAI_API_KEY env var.
            model: Model to use. Defaults to EMBEDDING_MODEL env var or provider default.
        """
        provider = provider or os.getenv("EMBEDDING_PROVIDER", "openai")
        self.provider_name = provider

        if provider == "openai":
            api_key = api_key or os.getenv("OPENAI_API_KEY")
            if not api_key:
                raise ValueError(
                    "OpenAI API key not provided. Set OPENAI_API_KEY environment variable "
                    "or pass api_key parameter."
                )
            model = model or os.getenv("EMBEDDING_MODEL", "text-embedding-3-small")
            self._provider = OpenAIProvider(api_key=api_key, model=model)

        elif provider == "local":
            model = model or os.getenv("EMBEDDING_MODEL", "BAAI/bge-small-en-v1.5")
            self._provider = LocalProvider(model=model)

        else:
            raise ValueError(f"Unknown embedding provider: {provider}")

        self.model = model
        logger.info(f"Initialized EmbeddingService with provider: {provider}, model: {model}")

    def clean_content(self, text: str) -> str:
        """
        Clean Silverbullet syntax noise from content before embedding.

        Removes:
        - Silverbullet attributes: #tag, @mention, etc.
        - Wikilinks: [[page]] or [[page|alias]]
        - Front matter delimiters
        - Excessive whitespace

        Args:
            text: Raw markdown content

        Returns:
            Cleaned text suitable for embedding
        """
        # Remove front matter delimiters
        text = re.sub(r'^---\s*$', '', text, flags=re.MULTILINE)

        # Convert wikilinks to plain text (keep the display text)
        # [[page|alias]] -> alias
        # [[page]] -> page
        text = re.sub(r'\[\[([^\]|]+)\|([^\]]+)\]\]', r'\2', text)
        text = re.sub(r'\[\[([^\]]+)\]\]', r'\1', text)

        # Remove Silverbullet attributes (tags, mentions, etc.)
        # Keep the text but remove the special syntax
        text = re.sub(r'#(\w+)', r'\1', text)  # #tag -> tag
        text = re.sub(r'@(\w+)', r'\1', text)  # @mention -> mention

        # Remove excessive whitespace
        text = re.sub(r'\n\s*\n\s*\n+', '\n\n', text)  # Multiple newlines -> double newline
        text = re.sub(r' +', ' ', text)  # Multiple spaces -> single space
        text = text.strip()

        return text

    def generate_embedding(self, text: str, clean: bool = True) -> List[float]:
        """
        Generate embedding for a single text.

        Args:
            text: Text to embed
            clean: Whether to clean Silverbullet syntax first (default: True)

        Returns:
            Embedding vector as list of floats
        """
        if clean:
            text = self.clean_content(text)

        if not text or not text.strip():
            logger.warning("Attempted to embed empty text, returning zero vector")
            return [0.0] * self._provider.get_dimension()

        try:
            embedding = self._provider.generate_embedding(text)
            logger.debug(f"Generated embedding with dimension: {len(embedding)}")
            return embedding
        except Exception as e:
            logger.error(f"Failed to generate embedding: {e}")
            raise

    def generate_embeddings_batch(
        self,
        texts: List[str],
        clean: bool = True
    ) -> List[List[float]]:
        """
        Generate embeddings for multiple texts in a batch.

        Args:
            texts: List of texts to embed
            clean: Whether to clean Silverbullet syntax first (default: True)

        Returns:
            List of embedding vectors
        """
        if clean:
            texts = [self.clean_content(text) for text in texts]

        # Filter out empty texts but keep track of indices
        valid_texts = []
        valid_indices = []
        for i, text in enumerate(texts):
            if text and text.strip():
                valid_texts.append(text)
                valid_indices.append(i)

        dimension = self._provider.get_dimension()

        if not valid_texts:
            logger.warning("All texts were empty, returning zero vectors")
            return [[0.0] * dimension] * len(texts)

        try:
            valid_embeddings = self._provider.generate_embeddings_batch(valid_texts)

            # Create result list with zero vectors for empty texts
            embeddings = [[0.0] * dimension] * len(texts)
            for i, embedding in enumerate(valid_embeddings):
                original_index = valid_indices[i]
                embeddings[original_index] = embedding

            logger.info(f"Generated {len(valid_texts)} embeddings in batch")
            return embeddings
        except Exception as e:
            logger.error(f"Failed to generate embeddings in batch: {e}")
            raise

    def get_embedding_dimension(self) -> int:
        """
        Get the dimension of embeddings for the current model.

        Returns:
            Embedding dimension (e.g., 1536 for OpenAI, 384 for bge-small)
        """
        return self._provider.get_dimension()

File: internal/embeddings/embeddings.go
// Package embeddings provides embedding generation for semantic search.
package embeddings

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/knights-analytics/hugot"
	"github.com/knights-analytics/hugot/pipelines"
)

// Provider represents an embedding provider type.
type Provider string

const (
	ProviderOpenAI Provider = "openai"
	ProviderLocal  Provider = "local"
)

// Local model constants
const (
	DefaultLocalModel = "BAAI/bge-small-en-v1.5"
	LocalModelDim     = 384
)

// Config holds embedding service configuration.
type Config struct {
	Provider  Provider
	APIKey    string
	Model     string
	BaseURL   string
	CacheDir  string // Directory to cache local models
	MaxLength int    // Max sequence length for local models
}

// Service generates text embeddings.
type Service struct {
	config   Config
	client   *http.Client
	session  *hugot.Session
	pipeline *pipelines.FeatureExtractionPipeline
	mu       sync.Mutex
}

// NewService creates a new embedding service.
func NewService(cfg Config) (*Service, error) {
	if cfg.Provider == "" {
		cfg.Provider = Provider(os.Getenv("EMBEDDING_PROVIDER"))
		if cfg.Provider == "" {
			cfg.Provider = ProviderOpenAI
		}
	}

	svc := &Service{
		config: cfg,
		client: &http.Client{Timeout: 30 * time.Second},
	}

	switch cfg.Provider {
	case ProviderOpenAI:
		if err := svc.initOpenAI(); err != nil {
			return nil, err
		}
	case ProviderLocal:
		if err := svc.initLocal(); err != nil {
			return nil, err
		}
	default:
		return nil, fmt.Errorf("unknown embedding provider: %s", cfg.Provider)
	}

	return svc, nil
}

func (s *Service) initOpenAI() error {
	if s.config.APIKey == "" {
		s.config.APIKey = os.Getenv("OPENAI_API_KEY")
	}
	if s.config.APIKey == "" {
		return fmt.Errorf("OpenAI API key not provided")
	}
	if s.config.Model == "" {
		s.config.Model = os.Getenv("EMBEDDING_MODEL")
		if s.config.Model == "" {
			s.config.Model = "text-embedding-3-small"
		}
	}
	if s.config.BaseURL == "" {
		s.config.BaseURL = "https://api.openai.com/v1"
	}
	return nil
}

func (s *Service) initLocal() error {
	if s.config.Model == "" {
		s.config.Model = DefaultLocalModel
	}
	if s.config.CacheDir == "" {
		s.config.CacheDir = filepath.Join(os.TempDir(), "hugot-models")
	}
	if s.config.MaxLength == 0 {
		s.config.MaxLength = 512
	}

	// Create hugot session with native Go backend
	session, err := hugot.NewGoSession()
	if err != nil {
		return fmt.Errorf("create hugot session: %w", err)
	}
	s.session = session

	// Download model if needed
	modelPath, err := hugot.DownloadModel(s.config.Model, s.config.CacheDir, hugot.NewDownloadOptions())
	if err != nil {
		_ = s.session.Destroy()
		return fmt.Errorf("download model %s: %w", s.config.Model, err)
	}

	// Create feature extraction pipeline
	pipelineConfig := hugot.FeatureExtractionConfig{
		ModelPath: modelPath,
		Name:      "embeddings",
	}
	pipeline, err := hugot.NewPipeline(s.session, pipelineConfig)
	if err != nil {
		_ = s.session.Destroy()
		return fmt.Errorf("create pipeline: %w", err)
	}
	s.pipeline = pipeline

	return nil
}

// Close releases resources used by the service.
func (s *Service) Close() error {
	if s.session != nil {
		_ = s.session.Destroy()
		s.session = nil
	}
	return nil
}

// CleanContent removes SilverBullet syntax noise from text.
func CleanContent(text string) string {
	// Remove front matter delimiters
	text = regexp.MustCompile(`(?m)^---\s*$`).ReplaceAllString(text, "")

	// Convert wikilinks: [[page|alias]] -> alias, [[page]] -> page
	text = regexp.MustCompile(`\[\[([^\]|]+)\|([^\]]+)\]\]`).ReplaceAllString(text, "$2")
	text = regexp.MustCompile(`\[\[([^\]]+)\]\]`).ReplaceAllString(text, "$1")

	// Remove SilverBullet attributes
	text = regexp.MustCompile(`#(\w+)`).ReplaceAllString(text, "$1")
	text = regexp.MustCompile(`@(\w+)`).ReplaceAllString(text, "$1")

	// Normalize whitespace
	text = regexp.MustCompile(`\n\s*\n\s*\n+`).ReplaceAllString(text, "\n\n")
	text = regexp.MustCompile(` +`).ReplaceAllString(text, " ")

	return strings.TrimSpace(text)
}

// GenerateEmbedding generates an embedding for a single text.
func (s *Service) GenerateEmbedding(ctx context.Context, text string, clean bool) ([]float32, error) {
	if clean {
		text = CleanContent(text)
	}

	if strings.TrimSpace(text) == "" {
		return make([]float32, s.GetDimension()), nil
	}

	embeddings, err := s.generateEmbeddings(ctx, []string{text})
	if err != nil {
		return nil, err
	}

	return embeddings[0], nil
}

// GenerateEmbeddingsBatch generates embeddings for multiple texts.
func (s *Service) GenerateEmbeddingsBatch(ctx context.Context, texts []string, clean bool) ([][]float32, error) {
	if clean {
		cleaned := make([]string, len(texts))
		for i, t := range texts {
			cleaned[i] = CleanContent(t)
		}
		texts = cleaned
	}

	// Track which texts are valid
	validTexts := make([]string, 0, len(texts))
	validIndices := make([]int, 0, len(texts))
	for i, t := range texts {
		if strings.TrimSpace(t) != "" {
			validTexts = append(validTexts, t)
			validIndices = append(validIndices, i)
		}
	}

	if len(validTexts) == 0 {
		result := make([][]float32, len(texts))
		dim := s.GetDimension()
		for i := range result {
			result[i] = make([]float32, dim)
		}
		return result, nil
	}

	validEmbeddings, err := s.generateEmbeddings(ctx, validTexts)
	if err != nil {
		return nil, err
	}

	// Build result with zero vectors for empty texts
	result := make([][]float32, len(texts))
	dim := s.GetDimension()
	for i := range result {
		result[i] = make([]float32, dim)
	}
	for i, embedding := range validEmbeddings {
		result[validIndices[i]] = embedding
	}

	return result, nil
}

// GetDimension returns the embedding dimension for the current model.
func (s *Service) GetDimension() int {
	if s.config.Provider == ProviderLocal {
		switch s.config.Model {
		case "BAAI/bge-small-en-v1.5", "BAAI/bge-small-en":
			return 384
		case "BAAI/bge-base-en-v1.5", "BAAI/bge-base-en":
			return 768
		case "sentence-transformers/all-MiniLM-L6-v2":
			return 384
		default:
			return LocalModelDim
		}
	}

	// OpenAI models
	switch s.config.Model {
	case "text-embedding-3-small":
		return 1536
	case "text-embedding-3-large":
		return 3072
	case "text-embedding-ada-002":
		return 1536
	default:
		return 1536
	}
}

// GetProvider returns the configured provider.
func (s *Service) GetProvider() Provider {
	return s.config.Provider
}

// GetModel returns the configured model name.
func (s *Service) GetModel() string {
	return s.config.Model
}

func (s *Service) generateEmbeddings(ctx context.Context, texts []string) ([][]float32, error) {
	switch s.config.Provider {
	case ProviderOpenAI:
		return s.generateOpenAIEmbeddings(ctx, texts)
	case ProviderLocal:
		return s.generateLocalEmbeddings(ctx, texts)
	default:
		return nil, fmt.Errorf("unknown provider: %s", s.config.Provider)
	}
}

func (s *Service) generateLocalEmbeddings(ctx context.Context, texts []string) ([][]float32, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.pipeline == nil {
		return nil, fmt.Errorf("local embedding pipeline not initialized")
	}

	// Run the pipeline
	batchResult, err := s.pipeline.RunPipeline(texts)
	if err != nil {
		return nil, fmt.Errorf("run pipeline: %w", err)
	}

	// FeatureExtractionOutput has Embeddings [][]float32
	return batchResult.Embeddings, nil
}

func (s *Service) generateOpenAIEmbeddings(ctx context.Context, texts []string) ([][]float32, error) {
	reqBody := map[string]any{
		"model": s.config.Model,
		"input": texts,
	}
	reqJSON, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", s.config.BaseURL+"/embeddings", bytes.NewReader(reqJSON))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.config.APIKey)

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var errResp struct {
			Error struct {
				Message string `json:"message"`
			} `json:"error"`
		}
		_ = json.NewDecoder(resp.Body).Decode(&errResp)
		return nil, fmt.Errorf("OpenAI API error (%d): %s", resp.StatusCode, errResp.Error.Message)
	}

	var result struct {
		Data []struct {
			Embedding []float32 `json:"embedding"`
			Index     int       `json:"index"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}

	// Sort by index to maintain order
	embeddings := make([][]float32, len(result.Data))
	for _, d := range result.Data {
		embeddings[d.Index] = d.Embedding
	}

	return embeddings, nil
}
